src/common/src/hash/key.rs

```diff
@@ -41,6 +41,39 @@ use crate::types::{DataType, Date, Decimal, ScalarRef, Time, Timestamp, F32, F64
 use crate::util::hash_util::Crc32FastBuilder;
 use crate::util::iter_util::ZipEqFast;
 use crate::util::value_encoding::{deserialize_datum, serialize_datum_into};
+use smallbitset::Set8;
+
+/// Bitmap for null values in key.
+/// This is specialized for key,
+/// since it usually has few group keys.
+#[repr(transparent)]
+#[derive(Clone, Debug, PartialEq)]
+pub struct NullBitmap {
+    inner: Set8,
```

Comment:
What if we have more than 8 keys? 👀

Reply:
Yeah this PR is still wip. Implemented and benchmarked case with 8 keys first. See comment above. That's still unsatisfactory, so I havent moved on to case with more than 8 keys.